### ISSUE TITLE
New compiler: treat empty args list after `new T` as non-error

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2174]
CC @fernewelten 

I cannot remember if there's a feature suggestion for type constructors in script; I would refer to that if it existed. But the current proposal is a slightly different thing.

I've been migrating a huge project from 3.6.0 to ags4 wanting to test how new compiler would work with it, and I found that the project's author was using a form `T* obj = new T();` in too many places. Either this is a habit from programming in C++/C# like languages, or copied from them. In any case, the old compiler seems to skip that.

My proposal is to teach new compiler to ignore the empty arg list after a type silently for the time being. Maybe this could be implemented as a sort of a *placeholder* for constructor feature, where only argument-less constructors are allowed.
This will make it much easier to migrate old scripts from ags3, or copy code from another programming languages where such syntax is common. Also, this might make it easier to transition to the new syntax if constructors will actually be supported at some point.

**Comment by fernewelten** [comment]
My suggestion would be:

When `S` is a `managed struct` that has a member function `init`, then you can do `new S(parameters)`. In this case, a new struct of type S is created - let's call it `s`- and afterwards, `s.init(parameters)` is called.

```
managed Struct Weapon
{
    int Resistance;
    int Durability
    import function init(int r, int d);
}
…
function game_start()
{
    Weapon *w = new Weapon(100,50);
}
```
In `game_start()`, w will be assigned a new struct of type `Weapon`, then `w.init(100,50)` will be called.

You can't go `new Weapon(2.0)` because the parameter list `(2.0)`doesn't match the declaration of `init`.

**Comment by fernewelten** [comment]
So as a first measure, I'm providing a PR so that the compiler accepts `()` after `new`. This parameter list is simply ignored for now, but a warning is issued.

**Comment by ivan-mogilko** [comment]
I'm sorry, but I disagree with 2 points from above.
First of all, having a difference between `new T` and `new T()` may be misleading and a source of mistakes.
Second, in the case described above, where user did not provide a "init" function, there will be too many warnings, which will clutter output and annoy users.

May I instead propose following:
1. No brackets and empty brackets is treated as equivalent, where "no brackets" is a old-style of the object creation:
```
// both cases call init if it's present
T* o1 = new T;
T* o2 = new T();
```
2. Empty brackets do not cause a warning if init is not found, in other words, compiler assumes there's a no-op init function with no args if one is not provided by a user.

**Comment by ivan-mogilko** [comment]
I have a slightly different proposal: in case `init` (or whatever it will be called) is not present, assume a no-op `init` by default.
This will means that `T* o = new T();` will not generate any warnings.

In my opinion this will be kinder to users, because with the #2181 compiler still may generate hundreds of warnings, eventually forcing user to edit the scripts.

**Comment by fernewelten** [comment]
> 1. No brackets and empty brackets is treated as equivalent, where "no brackets" is a "old-style" syntax:
> 
> ```
> // both cases call init with no args if it's present
> T* o1 = new T;
> T* o2 = new T();
> ```
> 
> 2. Empty brackets do not cause a warning if init is not found, in other words, compiler assumes there's a no-op init function with no args if one is not provided by a user.

I'm fine with that.

I can take the warning out of #2181; it's just a ‘Not yet done, but I'll keep going!’ message anyway :-)

**Comment by ivan-mogilko** [comment]
Ah, yes, thank you.

**Comment by ivan-mogilko** [comment]
Fixed by #2181.

The constructor suggestion should be open as a separate ticket perhaps.
